packages/frontend/workspace/src/providers/sync/peer.ts

```diff
@@ -0,0 +1,397 @@
+    while (abort.aborted === false) {
+      try {
+        await this.sync(abort);
+      } catch (err) {
+        if (err === 'manually-stop') {
+          return;
+        }
+
+        this.logger.error('sync error', err);
+      }
+      try {
+        this.logger.error('retry after 5 seconds');
+        this.status = SyncPeerStatus.Retrying;
+        await Promise.race([
+          new Promise<void>(resolve => {
+            setTimeout(resolve, 5 * 1000);
+          }),
+          new Promise((_, reject) => {
+            // exit if manually stopped
+            if (abort.aborted) {
+              reject(abort.reason);
+            }
+            abort.addEventListener('abort', () => {
+              reject(abort.reason);
+            });
+          }),
+        ]);
+      } catch (err) {
+        if (err === 'manually-stop') {
+          return;
+        }
+
+        // should never reach here
+        throw err;
+      }
+    }
+  }
+
+  private state: {
+    connectedDocs: Map<string, Doc>;
+    pushUpdatesQueue: AsyncQueue<{
+      docId: string;
+      data: Uint8Array;
+    }>;
+    pullUpdatesQueue: AsyncQueue<{
+      docId: string;
+      data: Uint8Array;
+    }>;
+    subdocsLoadQueue: AsyncQueue<Doc>;
+  } = {
+    connectedDocs: new Map(),
+    pushUpdatesQueue: new AsyncQueue(),
+    pullUpdatesQueue: new AsyncQueue(),
+    subdocsLoadQueue: new AsyncQueue(),
+  };
+
+  /**
+   * main synchronization logic
+   */
+  async sync(abortOuter: AbortSignal) {
+    const abortInner = new AbortController();
+
+    abortOuter.addEventListener('abort', reason => {
+      abortInner.abort(reason);
+    });
+
+    let dispose: (() => void) | null = null;
+    try {
+      // start listen storage updates
+      dispose = await this.storage.subscribe(
+        this.handleStorageUpdates,
+        reason => {
+          // abort if storage disconnect, should trigger retry loop
+          abortInner.abort('subscribe disconnect:' + reason);
+        }
+      );
+      throwIfAborted(abortInner.signal);
+
+      // Step 1: load root doc
+      this.status = SyncPeerStatus.LoadingRootDoc;
+
+      await this.connectDoc(this.rootDoc, abortInner.signal);
+
+      this.status = SyncPeerStatus.LoadingSubDoc;
+
+      // Step 2: load subdocs
+      this.state.subdocsLoadQueue.push(
+        ...Array.from(this.rootDoc.getSubdocs())
+      );
+
+      this.rootDoc.on('subdocs', this.handleSubdocsUpdate);
+
+      while (this.state.subdocsLoadQueue.length > 0) {
+        const subdoc = await this.state.subdocsLoadQueue.next(
+          abortInner.signal
+        );
+        await this.connectDoc(subdoc, abortInner.signal);
+      }
+
+      this.status = SyncPeerStatus.Syncing;
+      this.updateSyncStatus();
+
+      // Finally: start sync
+      await Promise.all([
+        // listen subdocs
+        (async () => {
+          while (throwIfAborted(abortInner.signal)) {
+            const subdoc = await this.state.subdocsLoadQueue.next(
```

Comment:
May also need to prevent page reload when these queues have pending items

Reply:
only local peer need to be cared

Reply:
yes, should be only the one that writes to persistent data sources

Reply:
Yes, i will do that after this PR